### PR TITLE
multiplexingDebugService for the web

### PR DESCRIPTION
**Pull request by sadasant**
Migrated the multiplexingDebugService to the web, for #9665 

This PR gets us closer to the meatier section of the migration of the variable views. The highlight in this case is making the `IJupyterDebugService` optional in the `multiplexingDebugService`.

Feedback appreciated!

### CODE REVIEW

File: src/platform/debugger/multiplexingDebugService.web.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (rchiodo) [inline src/platform/debugger/multiplexingDebugService.web.ts:1]: This seems too extreme to me. This code is a dupe of the other code. Can you not make a single file that just has the node specific parts optional?

**Review by rchiodo — CHANGES_REQUESTED** [review]
I think you could instead rename `multiplexingDebugService.node.ts` to `multiplexingDebugService.ts` and just have the node specific parts optional.

**Comment by sadasant** [comment]
@rchiodo can do! thank you for the feedback 👍

File: src/platform/debugger/multiplexingDebugService.ts [hunk 2]
```diff
@@ -78,21 +84,41 @@ export class MultiplexingDebugService implements IJupyterDebugService {
...
     }
     public registerDebugConfigurationProvider(debugType: string, provider: DebugConfigurationProvider): Disposable {
         const d1 = this.vscodeDebugService.registerDebugConfigurationProvider(debugType, provider);
-        const d2 = this.jupyterDebugService.registerDebugConfigurationProvider(debugType, provider);
-        return this.combineDisposables(d1, d2);
+        if (this.jupyterDebugService) {
+            const d2 = this.jupyterDebugService.registerDebugConfigurationProvider(debugType, provider);
+            return this.combineDisposables(d1, d2);
+        }
+        return {
```
Comment (rchiodo) [inline src/platform/debugger/multiplexingDebugService.ts:105]: I believe you can just return d1 in this case, not rewrite it to be an anonymous object that calls d1.dispose.

File: src/platform/debugger/multiplexingDebugService.ts [hunk 3]
```diff
@@ -78,21 +84,41 @@ export class MultiplexingDebugService implements IJupyterDebugService {
...
     }
     public registerDebugAdapterTrackerFactory(debugType: string, factory: DebugAdapterTrackerFactory): Disposable {
         const d1 = this.vscodeDebugService.registerDebugAdapterTrackerFactory(debugType, factory);
-        const d2 = this.jupyterDebugService.registerDebugAdapterTrackerFactory(debugType, factory);
-        return this.combineDisposables(d1, d2);
+        if (this.jupyterDebugService) {
+            const d2 = this.jupyterDebugService.registerDebugAdapterTrackerFactory(debugType, factory);
+            return this.combineDisposables(d1, d2);
+        }
+        return {
```
Comment (rchiodo) [inline src/platform/debugger/multiplexingDebugService.ts:117]: same here. Just return d1.

**Review by rchiodo — APPROVED** [review]
Looks good, just some minor fixups.

File: src/platform/debugger/multiplexingDebugService.ts [hunk 4]
```diff
@@ -78,21 +84,41 @@ export class MultiplexingDebugService implements IJupyterDebugService {
...
     public get onBreakpointHit(): Event<void> {
+        if (!this.jupyterDebugService) {
+            throw new Error('No jupyter debugger service');
+        }
         return this.jupyterDebugService.onBreakpointHit;
     }
     public startRunByLine(config: DebugConfiguration): Thenable<boolean> {
         this.lastStartedService = this.jupyterDebugService;
+        if (!this.jupyterDebugService) {
+            throw new Error('No run by line debugger service');
```
Comment (IanMatthewHuff) [inline src/platform/debugger/multiplexingDebugService.ts:95]: This is just a fault of our own naming issues. But I would probably just say "No Jupyter Debug Service" like you did on the previous Error just a few lines above. We inject it with the Run by line identifier, but it's really the Jupyter debug service missing. Or maybe "Couldn't start run by line. Jupyter Debug Service missing".

**Review by IanMatthewHuff — APPROVED** [review]

**Comment by sadasant** [comment]
@IanMatthewHuff @rchiodo thank you so much for your kind and thoughtful reviews! I will polish this up on Monday. 🙏 thanks again!

**Review by rchiodo — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**